python/ray/serve/controller.py

```diff
@@ -507,6 +507,12 @@ def deploy_apps(
 
         new_config_checkpoint = {}
 
+        existing_applications = set(
+            self.application_state_manager._application_states.keys()
+        )
+        new_applications = {app_config.name for app_config in config.applications}
+        self.delete_apps(existing_applications.difference(new_applications))
```

Comment:
Could you delete the old apps after all the new apps have deployed successfully? Otherwise, the cluster could get stuck in a weird intermediate state where it's failing to deploy new apps, but the old apps are also unavailable.

Reply:
Hm, I can move the deletion of old apps to after [the kickoff of the tasks for each application] (https://github.com/ray-project/ray/blob/bff750a95e8f51e7355822d701f2fe7b4fedce77/python/ray/serve/controller.py#L544-L553), but at that point there is still no guarantee that the the new apps have deployed successfully. Given that, what do you think are the benefits of moving the deletion after vs. before? Open to both approaches.

Reply:
For this point, can we move the deleting old app to the client side (client.py)? In this case, Shreyas point can be resolved.

Reply:
> For this point, can we move the deleting old app to the client side (client.py)? In this case, Shreyas point can be resolved.
In that case, client (and as a result serve agent) will be blocking. Discussed with @sihanwang41 offline and a better option if we wanted to guarantee successful deployment first is kick off a separate task that waits on all the other app deployment tasks. However, I'm not sure if this is the best choice.

> Could you delete the old apps after all the new apps have deployed successfully? Otherwise, the cluster could get stuck in a weird intermediate state where it's failing to deploy new apps, but the old apps are also unavailable.

This is a valid point. However I'm also concerned about another weird intermediate state where some of the apps have deployed successfully but not all of them; in that case, there is a mix of old apps and new apps running on the cluster. This can cause conflicts e.g. if a new app and an old app share a route prefix. Since we don't provide any downtime guarantees if they remove an app from the deploy config, I think it might be cleaner to delete all old apps first. @shrekris-anyscale @sihanwang41 WDYT?

Reply:
> However I'm also concerned about another weird intermediate state where some of the apps have deployed successfully but not all of them; in that case, there is a mix of old apps and new apps running on the cluster.

This is a good point. [We used to have automatic rollback](https://github.com/ray-project/ray/pull/22121) behavior to prevent these weird states, but we've removed it since rollback can be handled by the user or a higher-level service provider by reapplying an old config. Given this, I'm okay with not worrying too much about inconsistent intermediate states.

I still have slight preference for moving the deletion to the end since deleting deployments before the new app rolls out means there's a brief period where no valid app is running (and can serve requests). This is a minor concern though, since users should be deploying new apps on a separate cluster and switching over traffic if they want zero downtime.

Reply:
Updated!